iree/compiler/Conversion/LinalgToSPIRV/TileAndVectorizeInOneWorkgroupPass.cpp

```diff
@@ -542,64 +599,83 @@ void TileAndVectorizeInOneWorkgroupPass::runOnOperation() {
       applyCanonicalizationPatternsForTiling(context, funcOp);
 
       LLVM_DEBUG({
-        llvm::dbgs() << "--- After Promotion  ---\n";
+        llvm::dbgs() << "--- After workgroup memory promotion  ---\n";
         funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
         llvm::dbgs() << "\n\n";
       });
     }
 
+    // TODO(thomasraoux, antiagainst): Tiling to subgroups shouldn't be
+    // controlled by vectorization. This is needed due to historical reasons.
+    // Change the second level tiling to cyclic to loops and remove this.
     if (launchConfig.useVectorize()) {
-      {
-        OwningRewritePatternList secondLevelTilingPatterns(&getContext());
-        populateTilingToSubgroupPatterns(context, launchConfig,
-                                         secondLevelTilingPatterns);
-        (void)applyPatternsAndFoldGreedily(
-            funcOp, std::move(secondLevelTilingPatterns));
-        applyCanonicalizationPatternsForTiling(context, funcOp);
-        promoteSingleIterationLoops(funcOp);
+      OwningRewritePatternList secondLevelTilingPatterns(&getContext());
+      populateTilingToSubgroupPatterns(context, launchConfig,
+                                       secondLevelTilingPatterns);
+      (void)applyPatternsAndFoldGreedily(funcOp,
+                                         std::move(secondLevelTilingPatterns));
+      applyCanonicalizationPatternsForTiling(context, funcOp);
+      promoteSingleIterationLoops(funcOp);
 
-        LLVM_DEBUG({
-          llvm::dbgs() << "--- After Second level Tiling  ---\n";
-          funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
-          llvm::dbgs() << "\n\n";
-        });
-      }
+      LLVM_DEBUG({
+        llvm::dbgs() << "--- After tiling to subgroups ---\n";
+        funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+        llvm::dbgs() << "\n\n";
+      });
+    }
 
-      {
-        OwningRewritePatternList thirdLevelTilingPatterns(&getContext());
-        populateTilingToInvocationPatterns(context, launchConfig,
-                                           thirdLevelTilingPatterns);
-        (void)applyPatternsAndFoldGreedily(funcOp,
-                                           std::move(thirdLevelTilingPatterns));
-        applyCanonicalizationPatternsForTiling(context, funcOp);
-        promoteSingleIterationLoops(funcOp);
+    {
+      OwningRewritePatternList thirdLevelTilingPatterns(&getContext());
+      populateTilingToInvocationPatterns(context, launchConfig,
+                                         thirdLevelTilingPatterns);
+      (void)applyPatternsAndFoldGreedily(funcOp,
+                                         std::move(thirdLevelTilingPatterns));
+
+      // Remove trip-one loops created during cyclic loop distribution if we can
+      // prove the tiling was perfect.
+      RewritePatternSet canoncalizationPatterns(context);
+      populateAffineMinSCFCanonicalizationPattern(canoncalizationPatterns);
```

Comment:
Ideally we should be able to use this pattern from core 
`AffineMinRangeCanonicalizationPattern` instead that take the same range lambda. That would allow us to remove `populateAffineMinSCFCanonicalizationPattern` we have in MLIR eventually.

Reply:
Ah didn't know about that pattern.. I can take a look and do it later.